core/src/main/java/de/danoeh/antennapod/core/storage/DBWriter.java

```diff
@@ -118,30 +126,59 @@ private static boolean deleteFeedMediaSynchronous(
-            adapter.open();
-            adapter.setMedia(media);
-            adapter.close();
+        } else if (media.getFile_url().startsWith("content://")) {
+            // Local feed
 
-            if (media.getId() == PlaybackPreferences.getCurrentlyPlayingFeedMediaId()) {
-                PlaybackPreferences.writeNoMediaPlaying();
-                IntentUtils.sendLocalBroadcast(context, PlaybackServiceInterface.ACTION_SHUTDOWN_PLAYBACK_SERVICE);
+            DocumentFile documentFile = DocumentFile.fromSingleUri(
+                    context,
+                    Uri.parse(media.getFile_url())
+            );
 
-                NotificationManagerCompat nm = NotificationManagerCompat.from(context);
-                nm.cancel(R.id.notification_playing);
+            if (documentFile == null || !documentFile.exists() || !documentFile.delete()) {
+                MessageEvent evt = new MessageEvent(context.getString(R.string.delete_local_failed));
+                EventBus.getDefault().post(evt);
+                return false;
             }
 
-            // Gpodder: queue delete action for synchronization
-            FeedItem item = media.getItem();
-            EpisodeAction action = new EpisodeAction.Builder(item, EpisodeAction.DELETE)
-                    .currentTimestamp()
-                    .build();
-            SynchronizationQueueSink.enqueueEpisodeActionIfSynchronizationIsActive(context, action);
+            localDelete = true;
+        }
+
+        media.setDownloaded(false);
+        media.setFile_url(null);
+        media.setHasEmbeddedPicture(false);
+        PodDBAdapter adapter = PodDBAdapter.getInstance();
+        adapter.open();
+        adapter.setMedia(media);
+        adapter.close();
+
+        if (media.getId() == PlaybackPreferences.getCurrentlyPlayingFeedMediaId()) {
+            PlaybackPreferences.writeNoMediaPlaying();
+            IntentUtils.sendLocalBroadcast(context, PlaybackServiceInterface.ACTION_SHUTDOWN_PLAYBACK_SERVICE);
+
+            NotificationManagerCompat nm = NotificationManagerCompat.from(context);
+            nm.cancel(R.id.notification_playing);
+        }
+
+        // Gpodder: queue delete action for synchronization
+        FeedItem item = media.getItem();
+        EpisodeAction action = new EpisodeAction.Builder(item, EpisodeAction.DELETE)
+                .currentTimestamp()
+                .build();
+        SynchronizationQueueSink.enqueueEpisodeActionIfSynchronizationIsActive(context, action);
+
+        if (!localDelete) {
+            EventBus.getDefault().post(FeedItemEvent.updated(media.getItem()));
+        } else {
+            // FeedItemEvent only updates the state of an item, but with local feed, item is deleted completely.
+            // Instead we must do full update of this feed to get rid of the item
+            // Additionally, updateFeed() must be called from the Main thread to prevent a deadlock.
```

Comment:
This does IO on the main thread. That's not a good idea because refreshing local feeds can take quite long if there are new items that need to be scanned for chapters etc.

Reply:
Problem is that this call in the DB Writer is already on the `dbExec` thread. Calling it on the DB thread will eventually result in a deadlock when `DBWriter.setCompleteFeed().get()` is called. Future will never complete, because complete feed action is scheduled on the same thread that is waiting for that future to complete.

What do you suggest we do here? Do we just spawn a thread instead of going the main thread route?

Reply:
Maybe a private setCompleteFeedSynchronous method that is called both here and in the setCompleteFeed method?

Reply:
I would have to then make the entire copy of the `updateFeed` into `updateFeedSynchronous`, just to turn all calls into synchronous calls.

What I did instead is added a check to only exec on the DB thread, if caller is not already on the DB thread. Otherwise, it just executes synchronously. What do you think?

Reply:
My idea was to rename the method to "synchronous" and add an additional method that calls the synchronous version on the db thread. Then no copy is needed. However, I see you solved it differently. I will think about your solution (if I can come up with things that would break). One problem I see is that it currently might change the return values to Futures without a return value (but that should be easy to solve if that is even a problem). But generally, it might be the more elegant approach.